play-v27/src/main/scala/com/gu/googleauth/actions.scala

```diff
@@ -137,9 +137,20 @@ trait LoginSupport extends Logging {
         }
         logWarn(desc, e)
         Future.successful(redirectWithError(failureRedirectTarget, message))
+    }.flatMap { userIdentity =>
+      authConfig.twoFactorAuthChecker.map(requireTwoFactorAuthFor(userIdentity)).getOrElse(EitherT.pure(userIdentity))
     }
   }
 
+  private def requireTwoFactorAuthFor(userIdentity: UserIdentity)(checker: TwoFactorAuthChecker)(
+    implicit ec: ExecutionContext
+  ): EitherT[Future, Result, UserIdentity] = EitherT {
+    checker.check(userIdentity.email).map(userHas2FA => if (userHas2FA) Right(userIdentity) else {
+      logger.warn(s"failed-oauth-callback : user-does-not-have-2fa")
+      Left(redirectWithError(failureRedirectTarget, "You do not have 2FA enabled"))
```

Comment:
This message will be displayed to the user if they don't have 2FA - as in this Ophan PR [here](https://github.com/guardian/ophan/pull/5675):

![image](https://github.com/guardian/play-googleauth/assets/52038/8248dd58-5224-4fe6-9b66-0ebc83efc1f1)